homeassistant/components/light/zha.py

```diff
@@ -83,13 +110,81 @@ class Light(ZhaEntity, light.Light):
 
+    class OnOffListener:
+        """Listener for the OnOff Zigbee cluster."""
+
+        def __init__(self, entity):
+            """Initialize OnOffListener."""
+            self._entity = entity
+
+        def cluster_command(self, tsn, command_id, args):
+            """Handle commands received to this cluster."""
+            if command_id in (0x0000, 0x0040):
+                self._entity.set_state(False)
+            elif command_id in (0x0001, 0x0041, 0x0042):
+                self._entity.set_state(True)
+            elif command_id == 0x0002:
+                self._entity.set_state(not self._entity.is_on)
+
+        def attribute_updated(self, attrid, value):
+            """Handle attribute updates on this cluster."""
+            if attrid == 0:
+                self._entity.set_state(value)
+
+        def zdo_command(self, *args, **kwargs):
+            """Handle ZDO commands on this cluster."""
+            pass
+
+        def zha_send_event(self, cluster, command, args):
+            """Relay entity events to hass."""
+            pass  # don't let entities fire events
+
+    class LevelListener:
+        """Listener for the LevelControl Zigbee cluster."""
+
+        def __init__(self, entity):
+            """Initialize LevelListener."""
+            self._entity = entity
+
+        def cluster_command(self, tsn, command_id, args):
+            """Handle commands received to this cluster."""
+            if command_id in (0x0000, 0x0004):  # move_to_level, -with_on_off
+                self._entity.set_level(args[0])
+            elif command_id in (0x0001, 0x0005):  # move, -with_on_off
+                # We should dim slowly -- for now, just step once
+                rate = args[1]
+                if args[0] == 0xff:
+                    rate = 10  # Should read default move rate
+                self._entity.move_level(-rate if args[0] else rate)
+            elif command_id in (0x0002, 0x0006):  # step, -with_on_off
+                # Step (technically may change on/off)
+                self._entity.move_level(-args[1] if args[0] else args[1])
+
+        def attribute_updated(self, attrid, value):
+            """Handle attribute updates on this cluster."""
+            if attrid == 0:
+                self._entity._brightness = value
+
+        def zdo_command(self, *args, **kwargs):
+            """Handle ZDO commands on this cluster."""
+            pass
+
+        def zha_send_event(self, cluster, command, args):
+            """Relay entity events to hass."""
+            pass  # don't let entities fire events
+
     def __init__(self, **kwargs):
         """Initialize the ZHA light."""
         super().__init__(**kwargs)
         self._supported_features = 0
         self._color_temp = None
         self._hs_color = None
         self._brightness = None
+        from zigpy.zcl.clusters import general
+        self._in_listeners = {
+            general.OnOff.cluster_id: self.OnOffListener(self),
```

Comment:
If this is registering the update callback, it should be moved to `async_added_to_hass`.

Reply:
That’s just a dict that is used by the base entity, they are registered in async_added_to_hass in the base ZHA entity